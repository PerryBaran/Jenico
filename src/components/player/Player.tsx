import { useState, useEffect, Dispatch, SetStateAction, useRef } from 'react';
import style from './player.module.css';
import { SongInfo } from '../../Interface';
import Audio from './audio/Audio';
import Info from './info/Info';
import ProgresssBar from './progressbar/ProgresssBar';
import Time from './time/Time';
import MediaControls from './mediaControls/MediaControls';
import Volume from './volume/Volume';
import Tracklist from './tracklist/Tracklist';
import KeyboardListener from './keyboardListener/KeyboardListeners';

interface Props {
    data: SongInfo[], 
    playing: boolean, 
    setPlaying: Dispatch<SetStateAction<boolean>>, 
    songIndex: number, 
    setSongIndex: Dispatch<SetStateAction<number>>, 
    albumIndex: number, 
    setAlbumIndex: Dispatch<SetStateAction<number>>,
    formFocused: boolean
};

function Player(props: Props) {
    const {data, playing, setPlaying, songIndex, setSongIndex, albumIndex, setAlbumIndex, formFocused} = props;
    const [time, setTime] = useState(0);
    const [duration, setDuration] = useState(0);
    const [volume, setVolume] = useState(0.5);
    const audioRef = useRef<HTMLAudioElement | null>(null);

    useEffect(() => {
        const updateTimer = setInterval(() => {
            if (audioRef.current) {
                const seconds = Math.round(audioRef.current.currentTime)
                setTime(seconds)
            }    
        }, 250);
        return () => clearInterval(updateTimer);
    }, [time]);

    useEffect(() => {
        if (audioRef.current) {
            audioRef.current.volume = volume;
        }
    }, [volume, audioRef]);

    const skipSong = (forwards = true) => {
        const albumLength = data[albumIndex].songs.length;
        const playlistLength = data.length;
        if (forwards) {
            const tempSongIndex = songIndex + 1;
            if (tempSongIndex >= albumLength) {
                let tempAlbumIndex = albumIndex + 1;
                if (tempAlbumIndex >= playlistLength) {
                    setAlbumIndex(0);
                    setSongIndex(0);
                    return;
                } else {
                    setAlbumIndex(tempAlbumIndex)
                    setSongIndex(0);
                    return;
                };
            } else {
                setSongIndex(tempSongIndex);
                return;
            };
        } else {
            const tempSongIndex = songIndex - 1;
            if (tempSongIndex < 0) {
                let tempAlbumIndex = albumIndex - 1;
                if (tempAlbumIndex < 0) {
                    setAlbumIndex(playlistLength - 1);
                    setSongIndex(data[playlistLength - 1].songs.length - 1);
                    return;
                } else {
                    setAlbumIndex(tempAlbumIndex);
                    setSongIndex(data[albumIndex - 1].songs.length - 1);
                    return;
                };
            } else {
                setSongIndex(tempSongIndex);
                return;
            };
        };
    };

    return (
        <section className={style.container}>
            <Audio 
                data={data}
                playing={playing}
                songIndex={songIndex}
                albumIndex={albumIndex}
                audioRef={audioRef}
                setDuration={setDuration}
                skipSong={skipSong}/>
            <Info 
                data={data}
                songIndex={songIndex}
                albumIndex={albumIndex}/>
            <div className={style.center}>
                <ProgresssBar
                    audioRef={audioRef}
                    time={time}
                    duration={duration}/>
                <div className={style.belowBar}>
                    <Time time={time}/>
                    <MediaControls 
                        playing={playing}
                        setPlaying={setPlaying}
                        skipSong={skipSong}/>
                    <Time time={duration}/>
                </div>
            </div>
            <Volume
                volume={volume}
                setVolume={setVolume}/>
            <Tracklist 
                data={data}
                songIndex={songIndex}
                setSongIndex={setSongIndex}
                albumIndex={albumIndex}
                setAlbumIndex={setAlbumIndex}
                setPlaying={setPlaying}/>
            <KeyboardListener 
                playing={playing}
                setPlaying={setPlaying}
                volume={volume}
                setVolume={setVolume}
                skipSong={skipSong}
                formFocused={formFocused}/>
        </section>
    );
};

export default Player;