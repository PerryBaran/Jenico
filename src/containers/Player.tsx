import React from 'react';

function Player(props:any) {
    return (
        <div>
            
        </div>
    );
}

export default Player;