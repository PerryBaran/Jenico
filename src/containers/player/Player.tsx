import React, { useState, useEffect, Dispatch, SetStateAction, useRef, ChangeEvent } from 'react';
import style from './player.module.css';
import { SongInfo, Songs } from '../../Interface';
import { play, pause, skip, lowVolume, menu } from '../../media/icons/index';
import KeyboardListener from './KeyboardListeners';

function Player(props: {data: SongInfo[], playing: boolean, setPlaying: Dispatch<SetStateAction<boolean>>,songIndex: number, setSongIndex: Dispatch<SetStateAction<number>>, albumIndex: number, setAlbumIndex: Dispatch<SetStateAction<number>>}) {
    const {data, playing, setPlaying, songIndex, setSongIndex, albumIndex, setAlbumIndex} = props;
    const [volume, setVolume] = useState(0.5);
    const [pageIndex, setPageIndex] = useState(albumIndex)
    const [time, setTime] = useState('0:00');
    const [duration, setDuration] = useState('0:00');
    const audioRef = useRef<HTMLAudioElement | null>(null);
    const barRef = useRef<HTMLInputElement | null>(null);
    const volRef = useRef<HTMLInputElement | null>(null);

    useEffect(() => {
        if (audioRef.current) {
                if (playing) {
                audioRef.current.play();
            } else {
                audioRef.current.pause();
            }
        }
    }, [playing, songIndex]);

    useEffect(() => {
        setPageIndex(albumIndex)
    }, [albumIndex]);

    useEffect(() => {
        const updateTimer = setInterval(() => {
            if (audioRef.current && barRef.current) {
                const seconds = Math.round(audioRef.current.currentTime)
                if (!isNaN(seconds)) {
                    barRef.current.value = `${seconds}`
                    setTime(calcDisplayTime(seconds))
                }
            }    
        }, 250);
        return () => clearInterval(updateTimer);
    });

    useEffect(() => {
        if (audioRef.current) {
            audioRef.current.volume = volume;
        }
    }, [volume]);

    const calcDisplayTime = (seconds: number) => {
        let sec: number | string = seconds;
        let min = 0;
        while (sec >= 60) {
            min++;
            sec = sec - 60;
        }
        if (sec < 10) {
            sec = `0${sec}`;
        }
        return `${min}:${sec}`;
    };

    const changeTime = (value: number) => {
        if (audioRef.current) {
            audioRef.current.currentTime = value;
        }
    };

    const updateProgressBarDuration = () =>{
        if (audioRef.current && barRef.current) {
            const seconds = Math.round(audioRef.current.duration);
            barRef.current.max = `${seconds}`;
            setDuration(calcDisplayTime(seconds));
        }
    };

    const skipSong = (forwards = true) => {
        const albumLength = data[albumIndex].songs.length;
        const playlistLength = data.length;
        if (forwards) {
            const tempSongIndex = songIndex + 1;
            if (tempSongIndex >= albumLength) {
                let tempAlbumIndex = albumIndex + 1;
                if (tempAlbumIndex >= playlistLength) {
                    setAlbumIndex(0);
                    setSongIndex(0);
                    return;
                } else {
                    setAlbumIndex(tempAlbumIndex)
                    setSongIndex(0);
                    return;
                }
            } else {
                setSongIndex(tempSongIndex);
                return;
            }
        } else {
            const tempSongIndex = songIndex - 1;
            if (tempSongIndex < 0) {
                let tempAlbumIndex = albumIndex - 1;
                if (tempAlbumIndex < 0) {
                    setAlbumIndex(playlistLength - 1);
                    setSongIndex(data[playlistLength - 1].songs.length - 1);
                    return;
                } else {
                    setAlbumIndex(tempAlbumIndex);
                    setSongIndex(data[albumIndex - 1].songs.length - 1);
                    return;
                }
            } else {
                setSongIndex(tempSongIndex);
                return;
            }
        }
    };

    const changeVolume = (e: ChangeEvent<HTMLInputElement>) => {
        const currentVolume = Number(e.target.value)/100;
        setVolume(currentVolume);
    };

    const scrollTrackList = (forwards = true) => {
        const playlistLength = data.length;
        if (forwards) {
            const tempPageIndex = pageIndex + 1
            if (tempPageIndex >= playlistLength) {
                setPageIndex(0);
            } else {
                setPageIndex(tempPageIndex);
            }
        } else {
            const tempPageIndex = pageIndex - 1
            if (tempPageIndex < 0) {
                setPageIndex(playlistLength - 1);
            } else {
                setPageIndex(tempPageIndex);
            }
        }
    };

    const setSongTracklist = (index: number) => {
        setAlbumIndex(pageIndex);
        setSongIndex(index);
    }

    return (
        <section className={style.container}>
            <audio 
                ref={audioRef}
                src={data[albumIndex]?.songs[songIndex]?.src}
                onLoadedMetadata={() => {
                    updateProgressBarDuration();
                    playing && audioRef.current?.play();
                }}
                onEnded={() => skipSong()}/>
            <div className={style.start}>
                <img src={data[albumIndex]?.art} alt='album cover art'/>
                <div>
                    <h3>{data[albumIndex]?.title}</h3>
                    <p>{data[albumIndex]?.songs[songIndex]?.name}</p>
                </div>
            </div>
            <div className={style.center}>
                <div className={style.bar}>
                    <input 
                        type='range'
                        ref={barRef}
                        name='time'
                        min={0}
                        defaultValue={0}
                        onChange={e => changeTime(Number(e.target.value))}/>
                    <div className={style.belowBar}>
                        <div>
                            <p>{time}</p>
                        </div>
                        <div>
                            <div className={style.play}>
                                <button onClick={() => skipSong(false)}><img src={skip} alt='skip backward'/></button>
                                <button onClick={() => setPlaying(!playing)}><img src={playing? pause : play} alt='play/pause'/></button>
                                <button onClick={() => skipSong()}><img src={skip} alt='skip forward'/></button>
                            </div>
                        </div>
                        <div>
                            <p>{duration}</p>
                        </div>
                    </div>
                </div>
            </div>
            <div className={style.end}>
                <div className={style.tracklistContainer}>
                    <div className={style.tracklist}>
                        <div>
                            <button onClick={() => scrollTrackList(false)}>&lt;</button>
                            <h4>{data[pageIndex]?.title}</h4>
                            <button onClick={() => scrollTrackList()}>&gt;</button>
                        </div>
                        <ul>
                            {data[pageIndex]?.songs.map((song: Songs) => {
                                const index = data[pageIndex]?.songs.indexOf(song);
                                return (
                                    <li key={song.name}>
                                        <button className={style.tracklistButton} onClick={() => setSongTracklist(index)}>{song.name}</button>
                                    </li>
                                )
                            })}
                        </ul>
                    </div>
                    <button><img src={menu} alt='tracklist'/></button>
                </div>
                <div className={style.volume}>
                    <input
                        type='range'
                        name='volume'
                        ref={volRef}
                        min={0}
                        max={100}
                        defaultValue={volume * 100}
                        onChange={e => changeVolume(e)}/>
                    <button><img src={lowVolume} alt='volume'/></button>
                </div>
            </div>
            <KeyboardListener
                playing={playing}
                setPlaying={setPlaying}
                volume={volume}
                setVolume={setVolume}
                skipSong={skipSong}/>
        </section>
    );
}

export default Player;