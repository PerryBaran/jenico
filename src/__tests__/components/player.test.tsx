import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Player from "../../components/player/Player";
import mockData from "../../media/songInfo";

const ALBUM_INDEX = 1;

describe("Player", () => {
  const validProps = {
    data: mockData,
    playing: true,
    handlePlaying: jest.fn(),
    selectedSong: mockData[ALBUM_INDEX].songs[0].name,
    handleSelectedSong: jest.fn(),
    formFocused: true,
  };

  jest
    .spyOn(window.HTMLMediaElement.prototype, "play")
    .mockImplementation(jest.fn());
  jest
    .spyOn(window.HTMLMediaElement.prototype, "pause")
    .mockImplementation(jest.fn());

  afterEach(cleanup);

  test("snapshot", () => {
    const { asFragment } = render(
      <Player
        data={validProps.data}
        playing={validProps.playing}
        handlePlaying={validProps.handlePlaying}
        selectedSong={validProps.selectedSong}
        handleSelectedSong={validProps.handleSelectedSong}
        formFocused={validProps.formFocused}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("handleVolume", () => {
    render(
      <Player
        data={validProps.data}
        playing={validProps.playing}
        handlePlaying={validProps.handlePlaying}
        selectedSong={validProps.selectedSong}
        handleSelectedSong={validProps.handleSelectedSong}
        formFocused={validProps.formFocused}
      />
    );
    const input = screen.getByTestId("volume");

    expect(input).toHaveAttribute("value", "50");
    fireEvent.change(input, { target: { value: "200" } });
    expect(input).toHaveAttribute("value", "100");
    fireEvent.change(input, { target: { value: "-200" } });
    expect(input).toHaveAttribute("value", "0");
    fireEvent.change(input, { target: { value: "40" } });
    expect(input).toHaveAttribute("value", "40");
  });

  describe("skipSong", () => {
    describe("forwards", () => {
      test("nextSongIndex < albumLength", () => {
        const selectedSong = validProps.data[1].songs[0].name;
        render(
          <Player
            data={validProps.data}
            playing={validProps.playing}
            handlePlaying={validProps.handlePlaying}
            selectedSong={selectedSong}
            handleSelectedSong={validProps.handleSelectedSong}
            formFocused={validProps.formFocused}
          />
        );

        expect(validProps.handleSelectedSong).toBeCalledTimes(0);
        fireEvent.click(screen.getByAltText("skip forward"));
        expect(validProps.handleSelectedSong).toBeCalledTimes(1);
        expect(validProps.handleSelectedSong).toBeCalledWith(
          validProps.data[1].songs[1].name
        );
      });

      test("nextSongIndex > albumLength but nextAlbumIndex < playlistLength", () => {
        const selectedSong =
          validProps.data[1].songs[validProps.data[1].songs.length - 1].name;
        render(
          <Player
            data={validProps.data}
            playing={validProps.playing}
            handlePlaying={validProps.handlePlaying}
            selectedSong={selectedSong}
            handleSelectedSong={validProps.handleSelectedSong}
            formFocused={validProps.formFocused}
          />
        );

        expect(validProps.handleSelectedSong).toBeCalledTimes(0);
        fireEvent.click(screen.getByAltText("skip forward"));
        expect(validProps.handleSelectedSong).toBeCalledTimes(1);
        expect(validProps.handleSelectedSong).toBeCalledWith(
          validProps.data[2].songs[0].name
        );
      });

      test("nextSongIndex > albumLength and nextAlbumIndex > playlistLength", () => {
        const selectedSong =
          validProps.data[validProps.data.length - 1].songs[
            validProps.data[validProps.data.length - 1].songs.length - 1
          ].name;
        render(
          <Player
            data={validProps.data}
            playing={validProps.playing}
            handlePlaying={validProps.handlePlaying}
            selectedSong={selectedSong}
            handleSelectedSong={validProps.handleSelectedSong}
            formFocused={validProps.formFocused}
          />
        );

        expect(validProps.handleSelectedSong).toBeCalledTimes(0);
        fireEvent.click(screen.getByAltText("skip forward"));
        expect(validProps.handleSelectedSong).toBeCalledTimes(1);
        expect(validProps.handleSelectedSong).toBeCalledWith(
          validProps.data[0].songs[0].name
        );
      });
    });

    describe("backwards", () => {
      test("prevSongIndex >= 0", () => {
        const selectedSong = validProps.data[1].songs[1].name;
        render(
          <Player
            data={validProps.data}
            playing={validProps.playing}
            handlePlaying={validProps.handlePlaying}
            selectedSong={selectedSong}
            handleSelectedSong={validProps.handleSelectedSong}
            formFocused={validProps.formFocused}
          />
        );

        expect(validProps.handleSelectedSong).toBeCalledTimes(0);
        fireEvent.click(screen.getByAltText("skip backward"));
        expect(validProps.handleSelectedSong).toBeCalledTimes(1);
        expect(validProps.handleSelectedSong).toBeCalledWith(
          validProps.data[1].songs[0].name
        );
      });

      test("prevSongIndex < 0 but prevAlbumIndex >= 0", () => {
        const selectedSong = validProps.data[1].songs[0].name;
        render(
          <Player
            data={validProps.data}
            playing={validProps.playing}
            handlePlaying={validProps.handlePlaying}
            selectedSong={selectedSong}
            handleSelectedSong={validProps.handleSelectedSong}
            formFocused={validProps.formFocused}
          />
        );

        expect(validProps.handleSelectedSong).toBeCalledTimes(0);
        fireEvent.click(screen.getByAltText("skip backward"));
        expect(validProps.handleSelectedSong).toBeCalledTimes(1);
        expect(validProps.handleSelectedSong).toBeCalledWith(
          validProps.data[0].songs[validProps.data[0].songs.length - 1].name
        );
      });

      test("prevSongIndex < 0 and prevAlbumIndex < 0", () => {
        const selectedSong = validProps.data[0].songs[0].name;
        render(
          <Player
            data={validProps.data}
            playing={validProps.playing}
            handlePlaying={validProps.handlePlaying}
            selectedSong={selectedSong}
            handleSelectedSong={validProps.handleSelectedSong}
            formFocused={validProps.formFocused}
          />
        );

        expect(validProps.handleSelectedSong).toBeCalledTimes(0);
        fireEvent.click(screen.getByAltText("skip backward"));
        expect(validProps.handleSelectedSong).toBeCalledTimes(1);
        expect(validProps.handleSelectedSong).toBeCalledWith(
          validProps.data[validProps.data.length - 1].songs[
            validProps.data[validProps.data.length - 1].songs.length - 1
          ].name
        );
      });
    });
  });

  describe("keyDownHandler", () => {
    test("formFocus = true", () => {
      render(
        <Player
          data={validProps.data}
          playing={validProps.playing}
          handlePlaying={validProps.handlePlaying}
          selectedSong={validProps.selectedSong}
          handleSelectedSong={validProps.handleSelectedSong}
          formFocused={true}
        />
      );

      expect(validProps.handlePlaying).toBeCalledTimes(0);
      fireEvent.keyDown(window, { code: "Space", charCode: 32 });
      expect(validProps.handlePlaying).toBeCalledTimes(0);
    });

    test("formFocus = false", () => {
      render(
        <Player
          data={validProps.data}
          playing={validProps.playing}
          handlePlaying={validProps.handlePlaying}
          selectedSong={validProps.selectedSong}
          handleSelectedSong={validProps.handleSelectedSong}
          formFocused={false}
        />
      );
      const input = screen.getByTestId("volume");
      const initialVolume = "50";

      expect(validProps.handlePlaying).toBeCalledTimes(0);
      fireEvent.keyDown(window, { code: "Space", charCode: 32 });
      expect(validProps.handlePlaying).toBeCalledTimes(1);
      expect(validProps.handlePlaying).toBeCalledWith();
      expect(validProps.handleSelectedSong).toBeCalledTimes(0);
      fireEvent.keyDown(window, { code: "ArrowLeft", charCode: 37 });
      expect(validProps.handleSelectedSong).toBeCalledTimes(1);
      expect(validProps.handleSelectedSong).lastCalledWith(
        mockData[ALBUM_INDEX - 1].songs[
          mockData[ALBUM_INDEX - 1].songs.length - 1
        ].name
      );
      fireEvent.keyDown(window, { code: "ArrowRight", charCode: 39 });
      expect(validProps.handleSelectedSong).toBeCalledTimes(2);
      expect(validProps.handleSelectedSong).lastCalledWith(
        mockData[ALBUM_INDEX].songs[1].name
      );
      fireEvent.change(input, { target: { value: initialVolume } });
      expect(input).toHaveAttribute("value", initialVolume);
      fireEvent.keyDown(window, { code: "ArrowUp", charCode: 38 });
      expect(input).toHaveAttribute("value", `${Number(initialVolume) + 1}`);
      fireEvent.keyDown(window, { code: "Equal", charCode: 187 });
      expect(input).toHaveAttribute("value", `${Number(initialVolume) + 2}`);
      fireEvent.keyDown(window, { code: "NumpadAdd", charCode: 107 });
      expect(input).toHaveAttribute("value", `${Number(initialVolume) + 3}`);
      fireEvent.keyDown(window, { code: "ArrowDown", charCode: 40 });
      expect(input).toHaveAttribute("value", `${Number(initialVolume) + 2}`);
      fireEvent.keyDown(window, { code: "Minus", charCode: 189 });
      expect(input).toHaveAttribute("value", `${Number(initialVolume) + 1}`);
      fireEvent.keyDown(window, { code: "NumpadSubtract", charCode: 109 });
      expect(input).toHaveAttribute("value", initialVolume);
    });
  });
});
