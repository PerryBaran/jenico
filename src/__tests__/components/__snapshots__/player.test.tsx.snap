// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player snapshot 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <audio
      data-testid="audio"
    />
    <div
      class="container"
    >
      <img
        alt="Dreaming of Detuned Love cover art"
        src="DreamingOfDetunedLove.png"
      />
      <div>
        <h3>
          Dreaming of Detuned Love
        </h3>
        <p>
          Slumber
        </p>
      </div>
    </div>
    <div
      class="center"
    >
      <input
        class="bar"
        data-testid="progress-bar"
        max="0"
        min="0"
        name="time"
        type="range"
        value="0"
      />
      <div
        class="belowBar"
      >
        <div
          class="container"
        >
          <p>
            0:00
          </p>
        </div>
        <div
          class="container"
        >
          <button>
            <img
              alt="skip backward"
              src="skip.png"
            />
          </button>
          <button>
            <img
              alt="play/pause"
              src="pause.png"
            />
          </button>
          <button>
            <img
              alt="skip forward"
              src="skip.png"
            />
          </button>
        </div>
        <div
          class="container"
        >
          <p>
            0:00
          </p>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <input
        data-testid="volume"
        max="100"
        min="0"
        name="volume"
        type="range"
        value="50"
      />
      <img
        alt="volume"
        src="medium-volume.png"
      />
    </div>
    <div
      class="container"
    >
      <div
        class="tracklist"
      >
        <img
          alt="Dreaming of Detuned Love cover art"
          src="DreamingOfDetunedLove.png"
        />
        <div>
          <button>
            &lt;
          </button>
          <h4>
            Dreaming of Detuned Love
          </h4>
          <button>
            &gt;
          </button>
        </div>
        <ul>
          <li
            class="highlight"
          >
            <button>
              Slumber
            </button>
          </li>
          <li>
            <button>
              Your Careless Embrace
            </button>
          </li>
          <li>
            <button>
              Restless Thoughts
            </button>
          </li>
          <li>
            <button>
              Detuned Love
            </button>
          </li>
          <li>
            <button>
              Falling into the Void
            </button>
          </li>
        </ul>
      </div>
      <img
        alt="tracklist"
        class="tracklistMenu"
        src="menu.png"
      />
    </div>
  </section>
</DocumentFragment>
`;
